File: go/bob/bob.go
// Package bob is a lackadaisical teenbot
package bob

import (
	"strings"
	"unicode"
)

const answerYellQuestion = "Calm down, I know what I'm doing!"
const answerQuestion = "Sure."
const answerYell = "Whoa, chill out!"
const answerNothing = "Fine. Be that way!"
const answerAnything = "Whatever."

// Hey is your typical teen and returns
// a boring answer
func Hey(remark string) string {
	remark = strings.Trim(remark, " \t\n\r")

	if isEmpty(remark) {
		return answerNothing
	}

	if isQuestion(remark) {
		if isYelled(remark) {
			return answerYellQuestion
		}
		return answerQuestion
	}

	if isYelled(remark) {
		return answerYell
	}

	return answerAnything
}

func isYelled(remark string) bool {
	return strings.IndexFunc(remark, unicode.IsLetter) != -1 && remark == strings.ToUpper(remark)
}

func isQuestion(remark string) bool {
	return remark[len(remark)-1:] == "?"
}

func isEmpty(remark string) bool {
	return strings.Trim(remark, " \t") == ""
}

File: cpp/bob/bob.cpp
#include "bob.h"

#include <string>
#include <stdexcept>

namespace bob {
    bool is_question(std::string text) {
        for (int i=text.length()-1; i>=0; i--) {
            if (::isblank(text[i]) || ::isspace(text[i])) {
                continue;
            }
            if ('?' == text[i]) {
                return true;
            } else {
                break;
            }
        }
        return false;
    }

    bool is_upper_text(std::string text) {
        bool hasAlpha = false;
        for(const auto c: text) {
            if (::isalpha(c)) {
                if (::islower(c)) {
                    return false;
                }
                hasAlpha = true;
            }
        }
        return hasAlpha;
    }

    bool is_empty_text(std::string text) {
        for(const auto c: text) {
            if (!::isblank(c) && !::isspace(c)) {
                return false;
            }
        }
        return true;
    }

    std::string hey(std::string salute) {
        if (is_empty_text(salute)) {
            return "Fine. Be that way!";
        }
        if (is_question(salute)) {
            if (is_upper_text(salute)) {
                return "Calm down, I know what I'm doing!";
            }
            return "Sure.";
        }
        if (is_upper_text(salute)) {
            return "Whoa, chill out!";
        }

        return "Whatever.";
    }
}  // namespace bob
